src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.cs

```diff
@@ -36,13 +38,47 @@ internal abstract partial class AbstractDiagnosticsTaggerProvider<TTag> : Asynch
     {
         private readonly IDiagnosticService _diagnosticService;
 
+        /// <summary>
+        /// Keep track of the ITextSnapshot for the open Document that was used when diagnostics were
+        /// produced for it.  We need that because the DiagnoticService does not keep track of this
+        /// snapshot (so as to not hold onto a lot of memory), which means when we query it for 
+        /// diagnostics, we don't know how to map the span of the diagnostic to the current snapshot
+        /// we're tagging.
+        /// </summary>
+        private static readonly ConditionalWeakTable<object, ITextSnapshot> _diagnosticIdToTextSnapshot = 
+            new ConditionalWeakTable<object, ITextSnapshot>();
+
         protected AbstractDiagnosticsTaggerProvider(
             IDiagnosticService diagnosticService,
             IForegroundNotificationService notificationService,
             IAsynchronousOperationListener listener)
             : base(listener, notificationService)
         {
             _diagnosticService = diagnosticService;
+            _diagnosticService.DiagnosticsUpdated += OnDiagnosticsUpdated;
+        }
+
+        private void OnDiagnosticsUpdated(object sender, DiagnosticsUpdatedArgs e)
+        {
+            if (e.DocumentId == null)
+            {
+                return;
+            }
+
+            if (_diagnosticIdToTextSnapshot.TryGetValue(e.Id, out var snapshot))
+            {
+                return;
+            }
+
+            var document = e.Solution.GetDocument(e.DocumentId);
+            if (document != null && document.IsOpen() && document.TryGetText(out var sourceText))
```

Comment:
@jasonmalinowski TryGetText should pretty much always work for Documents that are open right?

I realize there is a slight race here in that IsOpen might return true, but then the file might lose, and TryGetText could fail.  but that's ok in terms of how this is coded up.

What i want to make sure is that in the case where a Document really is open, we should always feel confident hat TryGetText succeeds.  Is that something one can depend on?

Reply:
Personally, I'd delete those Try* methods before depending on them. :smile:

Reply:
Note: i could also just hold onto the Document.  The theory there being that since this is only for open Documents it's not expensive to hold onto.  While this could root some solutions, it would only be for a short while as the diagnostic subsystem would then recompute the diagnostics for those open documents quickly, replacing any exiting diagnostics.  however, it seems riskier.

@heejaechang what do you think?

Reply:
@CyrusNajmabadi Any reason not to just call GetText()? If your theory that TryGetText works is correct, it's instant. If not, well, you will still work?

Reply:
i can also just do a synchronous call to GetTextAsync().WaitAndGetResult().  But i feel *very* skeezy doing that.  i believe TryGetText should be safe, and i believe that's what rename uses and the same assumption it makes, right @jasonmalinowski ?

Reply:
Yeah, inline rename has this:

```c#
       private void InitializeOpenBuffers(SnapshotSpan triggerSpan)
        {
            using (Logger.LogBlock(FunctionId.Rename_CreateOpenTextBufferManagerForAllOpenDocs, CancellationToken.None))
            {
                HashSet<ITextBuffer> openBuffers = new HashSet<ITextBuffer>();
                foreach (var d in _workspace.GetOpenDocumentIds())
                {
                    var document = _baseSolution.GetDocument(d);
                    Contract.ThrowIfFalse(document.TryGetText(out var text));
```

So at least if you're on the UI thread, it's safe to assume an open document has a SourceText associated with it.

Reply:
I guess given right now there's a PR (#24693) out fixing that function from complexity like this, I'm not sure how much I'd trust the invariants of the system.

Reply:
Don't we have GetTextSynchronously() anyways? We have it for fetching trees, which implies we should be able to fetch text synchronously!

Reply:
> Any reason not to just call GetText()? If your theory that TryGetText works is correct, it's instant. If not, well, you will still work?

There is no GetText or GetTextSynchronously currently :(

Reply:
I can add that though if you think that's appropriate!

Reply:
SourceText for open document is referenced strongly in solution itself. so TryGetText should always return BUT!! Document.IsOpen() is mutable state not tied to Solution IsOpen is called off. so in other words, you could ask Solution that is created before Document is opened with Workspace which currently think Document is opened :)

so, TryGetText can return null... BUT! then text is most likely just came from closed file. so your fallback logic most likely give right text... :)

Reply:
Done!

Reply:
I just moved to GetTextSynchronously.